### ISSUE TITLE
CPU backup copies at ~2.4 GiB/s where a plain pinned torch copy on the same device does ~24 GiB/s

### ISSUE THREAD

**Issue by 0z5a** [fzyzcjy/torch_memory_saver#111]
## Summary

On an NVIDIA L20 (PCIe Gen4 x16), `pause()` with CPU backup moves 4 GiB at **2.41 GiB/s** and `resume()` at **5.32 GiB/s**, while a plain pinned `Tensor.copy_()` of the same volume on the same device in the same process does **24.43 GiB/s (D2H)** and **23.54 GiB/s (H2D)** — roughly 10x and 4.4x faster. With `TMS_INIT_ENABLE_CPU_BACKUP=0` the same `pause()` takes a few ms per GiB, so essentially all of the time is the backup copy rather than the VMM/allocator work around it.

This is not a bug report about correctness — the data is preserved and restored every time. It is a throughput question: is this the expected rate for the CPU-backup path, and is there a mode or configuration that gets closer to the device's copy bandwidth?

## Environment

| | |
|---|---|
| `torch_memory_saver` | `0.0.9.post1+slime` (installed from the slime-derived vime image; `torch_memory_saver_hook_mode_preload_cu13.abi3.so`) |
| GPU | NVIDIA L20, 44.4 GiB, PCIe Gen4 x16 |
| Driver / runtime | 570.86.10 / CUDA 13.0 |
| torch | 2.13.0+cu130 |
| Host | 503 GiB RAM, copies into pinned memory |

`nvidia-smi` reports the link trained to Gen4 x16 (`pcie.link.gen.current=4`, `pcie.link.width.current=16`) while the copies run, so the link is not sitting in a low-power state.

## Measurements

4 GiB bfloat16 tensor, single allocation, medians. Torch baselines use `copy_()` to/from a `pin_memory=True` and a pageable host buffer of the same size, in the same process:

| path | D2H | H2D |
|---|---|---|
| torch, pinned host buffer | 24.43 GiB/s | 23.54 GiB/s |
| torch, pageable host buffer | 8.75 GiB/s | 9.01 GiB/s |
| `tms.pause()` / `tms.resume()` | **2.41 GiB/s** | **5.32 GiB/s** |

Clean repeated run (`pause`/`resume` five times, draining the torch cache between iterations so each call moves exactly one 3.99 GiB tensor):

```
iter 0: pause=1644.1 ms resume=744.5 ms freed=3.97 GiB restored=3.99 GiB
iter 1: pause=1696.8 ms resume=793.5 ms freed=3.99 GiB restored=3.99 GiB
iter 2: pause=1658.1 ms resume=740.7 ms freed=3.99 GiB restored=3.99 GiB
iter 3: pause=1643.8 ms resume=752.1 ms freed=3.99 GiB restored=3.99 GiB
iter 4: pause=1666.6 ms resume=754.9 ms freed=3.99 GiB restored=3.99 GiB
median pause 1658.1 ms -> 2.41 GiB/s
median resume 752.1 ms -> 5.32 GiB/s
```

## CPU backup is the whole cost

Same `pause()`, same sizes, with the backup disabled so the memory is dropped instead of copied:

| size | pause, `TMS_INIT_ENABLE_CPU_BACKUP=1` | pause, `TMS_INIT_ENABLE_CPU_BACKUP=0` |
|---|---|---|
| 1 GiB | 424 ms | 2.0 ms |
| 2 GiB | 1231 ms | 4.8 ms |
| 4 GiB | 2842 ms | 7.7 ms |
| 8 GiB | 5992 ms | 30.0 ms |

(The backup-on column comes from a loop that reuses the process and accumulates earlier iterations' cached blocks, so it runs a little high; the clean single-tensor repeat above is the number to compare against.)

## What we ruled out

* **Number of allocations.** Holding the volume at 4 GiB and varying the live allocation count (1 / 64 / 512 / 4096) barely changes `pause()` for the first three (3568 / 3635 / 3236 ms); only 4096 x 1 MiB gets worse (8462 ms). A single 4 GiB allocation is already slow, so this is not per-copy fixed overhead dominating.
* **Link state / device.** Both torch baselines run in the same process, on the same device, at the same link generation, and reach ~24 GiB/s.
* **Pinned vs pageable host memory.** The comparison holds against a pageable buffer too (TMS is 3.6x slower than even that), and `cudaMallocHost` is in the binary, so the CPU side is presumably pinned already.
* **Disk backup.** `TMS_INIT_ENABLE_DISK_BACKUP=1` on ext4 is far worse (4 GiB: pause 49766 ms, resume 21052 ms), so it is not an alternative.

## Reproduction

```python
# repro.py
import statistics, time
import torch
from torch_memory_saver import torch_memory_saver as tms

GIB = 1024 ** 3
ELEMENTS = 4 * GIB // 2  # bfloat16


def timed(fn):
    start = time.perf_counter()
    fn()
    torch.cuda.synchronize()
    return time.perf_counter() - start


src = torch.empty(ELEMENTS, dtype=torch.bfloat16, device="cuda")
src.fill_(1)

host = torch.empty(ELEMENTS, dtype=torch.bfloat16, pin_memory=True)
d2h = statistics.median([timed(lambda: host.copy_(src)) for _ in range(3)])
h2d = statistics.median([timed(lambda: src.copy_(host)) for _ in range(3)])
print(f"torch pinned  D2H {4 / d2h:6.2f} GiB/s   H2D {4 / h2d:6.2f} GiB/s")
del host

pause = timed(tms.pause)
resume = timed(tms.resume)
print(f"tms           pause {4 / pause:6.2f} GiB/s ({pause * 1000:.0f} ms)   "
      f"resume {4 / resume:6.2f} GiB/s ({resume * 1000:.0f} ms)")
```

```bash
SO=$(python -c 'import os, torch_memory_saver; d = os.path.dirname(os.path.dirname(torch_memory_saver.__file__)); print(os.path.join(d, "torch_memory_saver_hook_mode_preload_cu13.abi3.so"))')
LD_PRELOAD="$SO" TMS_INIT_ENABLE=1 TMS_INIT_ENABLE_CPU_BACKUP=1 CUDA_VISIBLE_DEVICES=0 python repro.py
```

Output on the machine above:

```
torch pinned  D2H  24.43 GiB/s   H2D  23.54 GiB/s
tms           pause  2.41 GiB/s (1658 ms)   resume  5.32 GiB/s (752 ms)
```

## Questions

1. Is ~2.4 GiB/s the expected rate for the CPU-backup path, or is something falling back to a slow copy?
2. Would issuing the copies as `cudaMemcpyAsync` on a dedicated stream (rather than synchronous `cudaMemcpy` per allocation) recover part of the gap? The binary shows `cudaMemcpy` and `cudaDeviceSynchronize` in its imports.
3. Is there a backend or configuration other than CPU backup / disk backup that is expected to be faster for a ~18 GiB working set?

Context for why it matters: in a colocate RL loop the whole Megatron training stack is paused every step so vLLM can use the GPUs, ~18.7 GiB in our case, and this copy is 86% of the resulting `sleep()` + `wake_up()` cost — about 79% of a step on a small-model workload. Happy to run further measurements or test a patch on this machine if that would help.

**Comment by 0z5a** [comment]
Where this shows up, for context on impact: in [vime](https://github.com/vllm-project/vime)'s colocate RL loop the whole Megatron training stack is paused every step so vLLM can use the GPUs, ~18.7 GiB in our case (0.6B model; fp32 master weights, Adam moments and gradients dominate). This copy is 86% of the resulting `sleep()` + `wake_up()` cost and ~80% of a step on our small-model workload.

The measurement, the span breakdown and the probes behind the numbers above are in [vime#443](https://github.com/vllm-project/vime/pull/443), and the mode choice this affects is now annotated in the code in [vime#444](https://github.com/vllm-project/vime/pull/444).

**Comment by fzyzcjy** [comment]
feel free to PR, I think it should be able to achieve near normal tensor copy speed, and my first guess is via simple optimizations like using better kernels or whatever to parallelly copy etc
